### DISCUSSION TITLE
Funkcjonalności bazy danych

### DISCUSSION THREAD

**Discussion by ustankie** [ustankie/Trains#1] [Announcements]
# Główne funkcje systemu:
- wyświetlanie rozkładu jazdy danej trasy
- wyszukiwanie połączeń danego dnia/godziny wraz z limitem wyświetlonych wyników
- rezerwacja danego miejsca od stacji A do B
- wyświetlenie wolnych miejsc w danym połączeniu
- wyświetlenie wszystkich tras (cyklicznych) dla konkertnej instancji pociągu
- wyszukiwanie wszystkich rezerwacji użytkownika

# Schemat bazy danych

![image](https://github.com/ustankie/Trains/assets/117941711/fed2593c-b10e-4ab0-b40a-7aec3ffdf22d)

# Propozycja widoków, funkcji, procedur, triggerów
## Widoki
- [x] Wszystkie połączenia - K
- [x] listowanie dostępnych stacji (przydatne, by użytkownik mógł wybrać istniejące w bazie stacje początkowe i końcowe) - K

## Funkcje
- [x] wyszukiwanie wszystkich rezerwacji użytkownika - K
- wszystkie wolne miejsca na danym połączeniu od danej stacji do innej
- **wyspecjalizowanie tego powyżej, wyszukiwanie 1 losowego wolnego miejsca na danej relacji (szukanie wszystkich to niepotrzebny koszt obliczeniowy, nie widze zastosowania tego w naszej aplikacji)** - myślałam o tym, że klient może sobie wybrać dane miejsce
- wyszukiwanie miejsc danego typu (przy oknie/korytarzu/srodek )
- [x] wyszukanie wszystkich pasażerów na dane połączenie - K
- [x] Znajdowanie wszystkich tras przejezdzających z a do b danego dnia - K (ULA SPRAWDZ XD)
- [x] Zamiana nazwy stacji na ID stacji, w przypadku braku zwraca null - K
- [x] funkcja wyliczająca sumaryczną cenę za podróż od A do B, ze zniżką - U
- [x] funkcja zwracająca czas odjazdu z danej stacji - U 
- [x] funkcja zwracająca listę odcinków na trasie - U
- [x] funkcja zwracająca zajęte miejsca na danym odcinku - U
- [x] funkcja zwracająca wszystkie sections na trasie od A do B - U

## Procedury
- [x] dodanie użytkownika - U
-  [x] dodanie zniżki - U
- [x] dodanie siedzenia - U
- [x] dodanie nowej rezerwacji - U
- [x] zmiana statusu rezerwacji
- [x] procedura zwracająca błąd jeśli dane miejsce jest zajęte - U (zrealizowane w add_reservation)
- [ ] Usunięcie lub przywrócenie połączenia lub przywrócenie (pole active na false/true)
- [x] zmiana hasła użytkownika - K
- [x] dodanie połączenia - K, nie sprawdzam czy pociag jest wolny wtedy
- [x] dodanie stacji - K
- [ ] modyfikacja danych o odcinku
- [x] dodanie/usunięcie odcinka - U
- [x] dodanie pociagu - K 

## Triggery
- [x] logowanie dodania rezerwacji - U
- [ ] logowanie zmiany statusu
- [x] Nieopłacona rezerwacja staje się anulowana po 15 minutach
- [ ] trigger zabraniający usuwania logów z log_reservation
- [ ] trigger pilnujący zeby jeden pociag byl przypisany do jednej trasy w jednym oknie czasowym
- [ ] trigger na update statusu w reservations

## Constrainty
- status rezerwacji - możliwe wartości: N,P,C
- dzień tygodnia - możliwe wartości: 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'

## Checki
-  [ ] walidacja emaila, numeru telefonu - in progress

## Uwagi
- miejsca numerujemy tak jak PKP:
![image](https://github.com/ustankie/Trains/assets/117941711/c21db2c3-4da9-49b3-83ee-50b56664297a)

- tabela route_stations zawiera id stacji oraz id stacji, która występuje po niej na danej trasie.
- statusy rezerwacji: N - not payed, P - payed, C - cancelled, Nie można przejść z P albo C do N
- mysle ze dla uproszenia zakładamy ze pociag zaczyna dnia x i kończy też dnia x - good point

**Comment by ustankie** [comment]
# Aplikacja (backend, frontend)

W następnej kolejności na pewno trzeba zrobić rejestracje i logowanie, a dopiero potem rezerwacje (chyba że się zahardcoduje user id dodawane do bazy)
- [x] rejestracja
- [x] logowanie - to będzie trudne, będziemy potrzebować userContext dla całej apki, robiłam to już kiedyś ale nie pamiętam za dobrze
- [x] rezerwacje: zniżki, wyliczanie nowej ceny, dodawanie rezerwacji do bazy, rezerwacja siedzeń konkretnych
- [x] Wyświetlenie swoich rezerwacji

**Reply by krudny** [reply]
Tak, jeszcze miałem wizje na "wszystkie połączenia z danej stacji", ale nie wiem gdzie i jak to ładnie wkomponować
